api/pkg/transformer/feast/redis/instrumentation.go

```diff
@@ -0,0 +1,90 @@
+package redis
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/gojek/merlin/pkg/transformer"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	redisNewConn = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: transformer.PromNamespace,
+		Name:      "redis_new_connection_count",
+		Help:      "New connection established to redis server",
+	}, []string{})
+
+	redisPipeline = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: transformer.PromNamespace,
+		Name:      "redis_pipelined_count",
+		Help:      "Number of pipelined redis command",
+	}, []string{"status", "command"})
+
+	redisCommandLatency = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: transformer.PromNamespace,
+		Name:      "redis_command_latency_ms",
+		Help:      "Latency of redis command",
+		Buckets:   prometheus.ExponentialBuckets(1, 2, 10), // 1,2,4,8,16,32,64,128,256,512,+Inf
+	}, []string{"command"})
+
+	redisConnPoolStats = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: transformer.PromNamespace,
+		Name:      "redis_conn_stats",
+		Help:      "Redis connection stats",
+	}, []string{"stat"})
+
+	hitConnStats     = "hits"
+	missConnStats    = "misses"
+	timeoutConnStats = "timeouts"
+	idleConnStats    = "idle_conns"
+	staleConnStats   = "stale_conns"
+	totalConnStats   = "total_conns"
+
+	success = "success"
+	fail    = "fail"
+)
+
+type redisHook struct{}
+
+type requestStartKey struct{}
+
+func (h *redisHook) BeforeProcess(ctx context.Context, cmd redis.Cmder) (context.Context, error) {
+	return context.WithValue(ctx, requestStartKey{}, time.Now()), nil
+}
+
+func (h *redisHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	startTime, ok := ctx.Value(requestStartKey{}).(time.Time)
```

Comment:
How will it be able to calculate the latency? It doesn't seem to be the case here

Reply:
before running the redis command, it will create new context that add start time of the command, this part of `BeforeProcess`

Reply:
Ah got it!